browser-polyfill.js

```diff
@@ -0,0 +1,340 @@
+     *        of the wrapped method.
+     *
+     * @returns {Proxy<function>}
+     *        A Proxy object for the given method, which invokes the given wrapper
+     *        method in its place.
+     */
+    const wrapMethod = (target, method, wrapper) => {
+      return new Proxy(method, {
+        apply(targetMethod, thisObj, args) {
+          return wrapper.call(thisObj, target, ...args);
+        },
+      });
+    };
+
+    let hasOwnProperty = Function.call.bind(Object.prototype.hasOwnProperty);
+
+    /**
+     * Wraps an object in a Proxy which intercepts and wraps certain methods
+     * based on the given `wrappers` and `metadata` objects.
+     *
+     * @param {object} target
+     *        The target object to wrap.
+     *
+     * @param {object} [wrappers = {}]
+     *        An object tree containing wrapper functions for special cases. Any
+     *        function present in this object tree is called in place of the
+     *        method in the same location in the `target` object tree. These
+     *        wrapper methods are invoked as described in {@see wrapMethod}.
+     *
+     * @param {object} [metadata = {}]
+     *        An object tree containing metadata used to automatically generate
+     *        Promise-based wrapper functions for asynchronous. Any function in
+     *        the `target` object tree which has a corresponding metadata object
+     *        in the same location in the `metadata` tree is replaced with an
+     *        automatically-generated wrapper function, as described in
+     *        {@see wrapAsyncFunction}
+     *
+     * @returns {Proxy<object>}
+     */
+    const wrapObject = (target, wrappers = {}, metadata = {}) => {
+      let cache = Object.create(null);
+
+      let handlers = {
+        has(target, prop) {
+          return prop in target || prop in cache;
+        },
+
+        get(target, prop, receiver) {
+          if (prop in cache) {
+            return cache[prop];
+          }
+
+          if (!(prop in target)) {
+            return undefined;
+          }
+
+          let value = target[prop];
+
+          if (typeof value == "function") {
+            // This is a method on the underlying object. Check if we need to do
+            // any wrapping.
+
+            if (typeof wrappers[prop] === "function") {
+              // We have a special-case wrapper for this method.
+              value = wrapMethod(target, target[prop], wrappers[prop]);
+            } else if (hasOwnProperty(metadata, prop)) {
+              // This is an async method that we have metadata for. Create a
+              // Promise wrapper for it.
+              let wrapper = wrapAsyncFunction(prop, metadata[prop]);
+              value = wrapMethod(target, target[prop], wrapper);
+            } else {
+              // This is a method that we don't know or care about. Return the
+              // original method, bound to the underlying object.
+              value = value.bind(target);
```

Comment:
I think that it is better to not blindly forward methods or properties that you do not know. Otherwise it is not possible to support new promisified methods without breaking backwards-compatibility.